src/main/java/org/broadinstitute/hellbender/tools/spark/RevertSamSpark.java

```diff
@@ -0,0 +1,706 @@
+
+                // And then remove any tags that are calculated from the alignment
+                attrBroadcast.getValue().forEach(tag -> rec.setAttribute(tag, (String) null));
+                return rec;
+            });
+        }
+
+        return reads;
+    }
+
+    @VisibleForTesting
+    static Map<String, File> getOutputMap(
+            final File outputMapFile,
+            final File outputDir,
+            final String defaultExtension,
+            final List<SAMReadGroupRecord> readGroups,
+            final boolean outputByReadgroup) {
+
+        if (outputByReadgroup) {
+            final Map<String, File> outputMap;
+            if (outputMapFile != null) {
+                try {
+                    outputMap = createOutputMapFromFile(outputMapFile);
+                } catch (IOException e) {
+                    throw new GATKException("Encountered an error reading output map file", e);
+                }
+            } else {
+                outputMap = createOutputMapFromHeader(readGroups, outputDir, defaultExtension);
+            }
+            return outputMap;
+        } else {
+            return Collections.singletonMap(null, outputDir);
+        }
+    }
+
+    // Names the files based on the locations laid out in the readgroup map
+    private static Map<String, File> createOutputMapFromFile(final File outputMapFile) throws IOException {
+        final Map<String, File> outputMap = new HashMap<>();
+        final FeatureReader<TableFeature>  parser = AbstractFeatureReader.getFeatureReader(outputMapFile.getAbsolutePath(), new TableCodec(null), false);
+        for (final TableFeature row : parser.iterator()) {
+            final String id = row.get("READ_GROUP_ID");
+            final String output = row.get("OUTPUT");
+            final File outputPath = new File(output);
+            outputMap.put(id, outputPath);
+        }
+        CloserUtil.close(parser);
+        return outputMap;
+    }
+
+    // Names the files based on the readgroups individually presented in the header
+    private static Map<String, File> createOutputMapFromHeader(final List<SAMReadGroupRecord> readGroups, final File outputDir, final String extension) {
+        final Map<String, File> outputMap = new HashMap<>();
+        for (final SAMReadGroupRecord readGroup : readGroups) {
+            final String id = readGroup.getId();
+            final String fileName = id + extension;
+            final Path outputPath = Paths.get(outputDir.toString(), fileName);
+            outputMap.put(id, outputPath.toFile());
+        }
+        return outputMap;
+    }
+
+// ________________________________________________________________________________________________________________________
+// sum garbage
+// ________________________________________________________________________________________________________________________
+    /**
+     * Methods used for validating parameters to RevertSam.
+     */
+    static class ValidationUtil {
```

Comment:
lets inline this class unless there's a really good reason to have it separate...